File: app/src/main/java/io/github/rosemoe/sora/langs/textmate/VCSpaceTMLanguage.java
package io.github.rosemoe.sora.langs.textmate;

import android.os.Bundle;
import androidx.annotation.NonNull;
import com.raredev.vcspace.editor.completion.SimpleCompletionIconDrawer;
import com.raredev.vcspace.editor.completion.SimpleCompletionItemKind;
import com.raredev.vcspace.editor.completion.SimpleSnippetCompletionItem;
import com.raredev.vcspace.plugin.Plugin;
import com.raredev.vcspace.plugin.PluginsLoader;
import com.raredev.vcspace.util.ILogger;
import com.raredev.vcspace.util.PreferencesUtils;
import io.github.rosemoe.sora.lang.completion.CompletionHelper;
import io.github.rosemoe.sora.lang.completion.CompletionPublisher;
import io.github.rosemoe.sora.lang.completion.SnippetDescription;
import io.github.rosemoe.sora.lang.completion.snippet.parser.CodeSnippetParser;
import io.github.rosemoe.sora.lang.format.Formatter;
import io.github.rosemoe.sora.langs.textmate.registry.GrammarRegistry;
import io.github.rosemoe.sora.langs.textmate.registry.ThemeRegistry;
import io.github.rosemoe.sora.text.CharPosition;
import io.github.rosemoe.sora.text.Content;
import io.github.rosemoe.sora.text.ContentReference;
import io.github.rosemoe.sora.text.TextRange;
import io.github.rosemoe.sora.util.MyCharacter;
import io.github.rosemoe.sora.widget.SymbolPairMatch;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.eclipse.tm4e.core.grammar.IGrammar;
import org.eclipse.tm4e.languageconfiguration.model.AutoClosingPairConditional;
import org.eclipse.tm4e.languageconfiguration.model.LanguageConfiguration;

public class VCSpaceTMLanguage extends TextMateLanguage {

  protected TMFormatter formatter;
  protected final String languageScope;

  private List<SimplePluginCompletionItem> pluginsCompletion;

  protected VCSpaceTMLanguage(
      IGrammar grammar,
      LanguageConfiguration languageConfiguration,
      ThemeRegistry themeRegistry,
      String languageScope) {
    super(grammar, languageConfiguration, null, themeRegistry, false);
    this.languageScope = languageScope;

    pluginsCompletion = new ArrayList<>();
    readPluginSnippets();
  }

  public static VCSpaceTMLanguage create(String languageScopeName) {
    final GrammarRegistry grammarRegistry = GrammarRegistry.getInstance();
    var grammar = grammarRegistry.findGrammar(languageScopeName);

    if (grammar == null) {
      throw new IllegalArgumentException(
          String.format("Language with %s scope name not found", grammarRegistry));
    }

    var languageConfiguration = grammarRegistry.findLanguageConfiguration(grammar.getScopeName());

    return new VCSpaceTMLanguage(
        grammar, languageConfiguration, ThemeRegistry.getInstance(), languageScopeName);
  }

  @Override
  public void requireAutoComplete(
      @NonNull ContentReference content,
      @NonNull CharPosition position,
      @NonNull CompletionPublisher publisher,
      @NonNull Bundle extraArguments) {
    var prefix =
        CompletionHelper.computePrefix(content, position, MyCharacter::isJavaIdentifierPart);

    for (var pluginSnippet : pluginsCompletion) {
      if (pluginSnippet.label.startsWith(prefix) && prefix.length() > 0) {
        publisher.addItem(
            new SimpleSnippetCompletionItem(
                pluginSnippet.label, /* Label */
                pluginSnippet.desc, /* Desc */
                pluginSnippet.type, /* Type */
                SimpleCompletionIconDrawer.draw(SimpleCompletionItemKind.SNIPPET),
                new SnippetDescription(
                    prefix.length(), CodeSnippetParser.parse(pluginSnippet.snippet), true)));
      }
    }
  }

  @NonNull
  @Override
  public Formatter getFormatter() {
    if (formatter == null) {
      formatter = new TMFormatter(this);
    }
    return formatter;
  }

  @Override
  public int getTabSize() {
    return PreferencesUtils.getEditorTABSize();
  }

  @Override
  public boolean useTab() {
    return !PreferencesUtils.useSpaces();
  }

  @Override
  public SymbolPairMatch getSymbolPairs() {
    SymbolPairMatch symbolPair = new SymbolPairMatch();
    if (languageConfiguration == null) {
      return symbolPair;
    }

    List<AutoClosingPairConditional> autoClosingPairs = languageConfiguration.getAutoClosingPairs();
    if (autoClosingPairs == null) {
      return symbolPair;
    }
    for (AutoClosingPairConditional autoClosingPair : autoClosingPairs) {
      symbolPair.putPair(
          autoClosingPair.open,
          new SymbolPairMatch.SymbolPair(
              autoClosingPair.open,
              autoClosingPair.close,
              new TextMateSymbolPairMatch.SymbolPairEx(autoClosingPair)));
    }
    return symbolPair;
  }

  public String formatCode(Content text, TextRange range) {
    return text.toString();
  }

  public LanguageConfiguration getLanguageConfiguration() {
    return this.languageConfiguration;
  }

  public void editorCommitText(CharSequence text) {}

  public boolean checkIsCompletionChar(char c) {
    return MyCharacter.isJavaIdentifierPart(c);
  }

  // Load Snippets from plugins
  public void readPluginSnippets() {
    var plugins = PluginsLoader.plugins;

    for (Map.Entry<String, Plugin> entry : plugins.entrySet()) {
      var plugin = entry.getValue();

      if (plugin.snippet != null && plugin.snippet.getLanguageScope().equals(languageScope)) {
        var snippetFile = new File(entry.getKey() + "/" + plugin.snippet.getSnippetFilePath());

        if (snippetFile.exists()) {
          try {
            BufferedReader reader = new BufferedReader(new FileReader(snippetFile));
            String line;
            while ((line = reader.readLine()) != null) {
              String[] parts = line.split("::");
              if (parts.length == 4) {
                pluginsCompletion.add(
                    new SimplePluginCompletionItem(
                        parts[0], /* Label */
                        parts[1], /* Desc */
                        parts[2], /* Type */
                        parts[3] /* Snippet */));
              }
            }
            reader.close();
          } catch (IOException e) {
            ILogger.error("VCSpaceTMLanguage", e);
            e.printStackTrace();
          }
        }
      }
    }
  }

  private class SimplePluginCompletionItem {
    public String label;
    public String desc;
    public String type;
    public String snippet;

    public SimplePluginCompletionItem(String label, String desc, String type, String snippet) {
      this.label = label;
      this.desc = desc;
      this.type = type;
      this.snippet = snippet.replaceAll("\n", System.getProperty("line.separator"));
    }
  }
}

File: app/src/main/java/com/teixeira/vcspace/editor/language/textmate/VCSpaceTMLanguage.kt
/*
 * This file is part of Visual Code Space.
 *
 * Visual Code Space is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Visual Code Space is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Visual Code Space.
 * If not, see <https://www.gnu.org/licenses/>.
 */
package com.teixeira.vcspace.editor.language.textmate

import android.os.Bundle
import androidx.annotation.WorkerThread
import io.github.rosemoe.sora.lang.EmptyLanguage
import io.github.rosemoe.sora.lang.analysis.AnalyzeManager
import io.github.rosemoe.sora.lang.completion.CompletionHelper
import io.github.rosemoe.sora.lang.completion.CompletionPublisher
import io.github.rosemoe.sora.lang.completion.IdentifierAutoComplete
import io.github.rosemoe.sora.langs.textmate.registry.GrammarRegistry
import io.github.rosemoe.sora.langs.textmate.registry.ThemeRegistry
import io.github.rosemoe.sora.langs.textmate.registry.model.DefaultGrammarDefinition
import io.github.rosemoe.sora.langs.textmate.registry.model.GrammarDefinition
import io.github.rosemoe.sora.langs.textmate.utils.StringUtils
import io.github.rosemoe.sora.text.CharPosition
import io.github.rosemoe.sora.text.ContentReference
import io.github.rosemoe.sora.util.MyCharacter
import kotlinx.coroutines.runBlocking
import org.eclipse.tm4e.core.grammar.IGrammar
import org.eclipse.tm4e.core.registry.IGrammarSource
import org.eclipse.tm4e.core.registry.IThemeSource
import org.eclipse.tm4e.languageconfiguration.internal.model.LanguageConfiguration
import java.io.Reader

class VCSpaceTMLanguage protected constructor(
    val grammar: IGrammar?,
    @JvmField
    var languageConfiguration: LanguageConfiguration?,
    var grammarRegistry: GrammarRegistry,
    var themeRegistry: ThemeRegistry,
    @JvmField val createIdentifiers: Boolean
) : EmptyLanguage() {
    /**
     * Set tab size. The tab size is used to compute code blocks.
     */
    @JvmField
    var tabSize: Int = 4

    private var useTab = false

    val autoCompleter: IdentifierAutoComplete = IdentifierAutoComplete()
    var isAutoCompleteEnabled: Boolean = true

    var textMateAnalyzer: VCSpaceTMAnalyzer? = null

    private lateinit var newlineHandlers: Array<VCSpaceTMNewlineHandler>

    // this.grammar = grammar;
    var symbolPairMatch: VCSpaceTMSymbolPairMatch = VCSpaceTMSymbolPairMatch(this)

    var newlineHandler: VCSpaceTMNewlineHandler? = null
        private set

    init {
        createAnalyzerAndNewlineHandler(grammar, languageConfiguration)
    }


    /**
     * When you update the [TextMateColorScheme] for editor, you need to synchronize the updates here
     *
     * @param theme IThemeSource creates from file
     */
    @WorkerThread
    @Deprecated(
        "Use {@link ThemeRegistry#setTheme(String)}",
        ReplaceWith("themeRegistry.loadTheme(theme)")
    )
    @Throws(
        Exception::class
    )
    fun updateTheme(theme: IThemeSource?) {
        //if (textMateAnalyzer != null) {
        //  textMateAnalyzer.updateTheme(theme);
        //}
        themeRegistry.loadTheme(theme)
    }


    private fun createAnalyzerAndNewlineHandler(
        grammar: IGrammar?,
        languageConfiguration: LanguageConfiguration?
    ) {
        val old = textMateAnalyzer
        if (old != null) {
            old.receiver = null
            old.destroy()
        }
        try {
            textMateAnalyzer =
                VCSpaceTMAnalyzer(
                    this,
                    grammar,
                    languageConfiguration,  /*grammarRegistry,*/
                    themeRegistry
                )
        } catch (e: Exception) {
            e.printStackTrace()
        }
        this.languageConfiguration = languageConfiguration
        newlineHandler = VCSpaceTMNewlineHandler(this)
        newlineHandlers = arrayOf(newlineHandler!!)
        if (languageConfiguration != null) {
            // because the editor will only get the symbol pair matcher once
            // (caching object to stop repeated new object created),
            // the symbol pair needs to be updated inside the symbol pair matcher.
            symbolPairMatch.updatePair()
        }
    }

    fun updateLanguage(scopeName: String?) {
        val grammar = grammarRegistry.findGrammar(scopeName)
        val languageConfiguration = grammarRegistry.findLanguageConfiguration(
            grammar!!.scopeName
        )
        createAnalyzerAndNewlineHandler(grammar, languageConfiguration)
    }

    fun updateLanguage(grammarDefinition: GrammarDefinition?) {
        val grammar = grammarRegistry.loadGrammar(grammarDefinition)

        val languageConfiguration = grammarRegistry.findLanguageConfiguration(grammar.scopeName)

        createAnalyzerAndNewlineHandler(grammar, languageConfiguration)
    }

    override fun getAnalyzeManager(): AnalyzeManager {
        if (textMateAnalyzer == null) {
            return EmptyAnalyzeManager.INSTANCE
        }
        return textMateAnalyzer as VCSpaceTMAnalyzer
    }


    override fun useTab(): Boolean {
        return useTab
    }

    fun useTab(useTab: Boolean) {
        this.useTab = useTab
    }

    override fun getSymbolPairs(): VCSpaceTMSymbolPairMatch {
        return symbolPairMatch
    }

    override fun getNewlineHandlers(): Array<VCSpaceTMNewlineHandler> {
        return newlineHandlers
    }

    override fun requireAutoComplete(
        content: ContentReference,
        position: CharPosition,
        publisher: CompletionPublisher,
        extraArguments: Bundle
    ) {
        if (!isAutoCompleteEnabled) {
            return
        }
        val prefix =
            CompletionHelper.computePrefix(content, position, MyCharacter::isJavaIdentifierPart)

        val ref = content.reference
        val cursor = ref.cursor

        if (prefix.isNotEmpty()) {
            runBlocking {
                val idt = textMateAnalyzer!!.syncIdentifiers
                autoCompleter.requireAutoComplete(content, position, prefix, publisher, idt)

                grammar?.name?.let { grammarName ->
                    println(grammarName)
//          Gemini.completeCode(
//            CompletionMetadata(
//              language = grammarName,
//              textBeforeCursor = ref.subContent(
//                1,
//                1,
//                cursor.leftLine,
//                cursor.leftColumn
//              ).toString(),
//              textAfterCursor = ref.subContent(
//                cursor.leftLine,
//                cursor.leftColumn,
//                ref.lineCount - 1,
//                ref.getColumnCount(ref.lineCount - 1)
//              ).toString()
//            )
//          ).onSuccess {
//            val response = Gemini.removeBackticksFromMarkdownCodeBlock(
//              it.candidates[0].content.parts[0].asTextOrNull()
//            )
//            println(response)
//            withContext(Dispatchers.Main.immediate) {
//              publisher.addItem(
//                AICompletionItem(
//                  label = "$prefix${response.trim()}",
//                  desc = "AI Generated",
//                  commitText = response
//                )
//              )
//            }
//          }.onFailure(Throwable::printStackTrace)
                }
            }
        }
    }

    fun setCompleterKeywords(keywords: Array<String?>?) {
        autoCompleter.setKeywords(keywords, false)
    }

    companion object {
        @Deprecated("")
        fun prepareLoad(
            grammarSource: IGrammarSource,
            languageConfiguration: Reader?,
            themeSource: IThemeSource?
        ): IGrammar {
            val definition = DefaultGrammarDefinition.withGrammarSource(
                grammarSource,
                StringUtils.getFileNameWithoutExtension(grammarSource.filePath),
                null
            )
            val languageRegistry = GrammarRegistry.getInstance()
            val grammar = languageRegistry.loadGrammar(definition)
            if (languageConfiguration != null) {
                languageRegistry.languageConfigurationToGrammar(
                    LanguageConfiguration.load(
                        languageConfiguration
                    ), grammar
                )
            }
            val themeRegistry = ThemeRegistry.getInstance()
            try {
                themeRegistry.loadTheme(themeSource)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return grammar
        }

        @Deprecated("")
        fun create(
            grammarSource: IGrammarSource,
            languageConfiguration: Reader?,
            themeSource: IThemeSource?
        ): VCSpaceTMLanguage {
            val grammar = prepareLoad(grammarSource, languageConfiguration, themeSource)
            return create(grammar.scopeName, true)
        }

        @Deprecated("")
        fun create(grammarSource: IGrammarSource, themeSource: IThemeSource?): VCSpaceTMLanguage {
            val grammar = prepareLoad(grammarSource, null, themeSource)
            return create(grammar.scopeName, true)
        }

        @Deprecated("")
        fun createNoCompletion(
            grammarSource: IGrammarSource,
            languageConfiguration: Reader?,
            themeSource: IThemeSource?
        ): VCSpaceTMLanguage {
            val grammar = prepareLoad(grammarSource, languageConfiguration, themeSource)
            return create(grammar.scopeName, false)
        }

        @Deprecated("")
        fun createNoCompletion(
            grammarSource: IGrammarSource,
            themeSource: IThemeSource?
        ): VCSpaceTMLanguage {
            val grammar = prepareLoad(grammarSource, null, themeSource)
            return create(grammar.scopeName, false)
        }

        fun create(languageScopeName: String?, autoCompleteEnabled: Boolean): VCSpaceTMLanguage {
            return create(languageScopeName, GrammarRegistry.getInstance(), autoCompleteEnabled)
        }

        fun create(
            languageScopeName: String?,
            grammarRegistry: GrammarRegistry,
            autoCompleteEnabled: Boolean
        ): VCSpaceTMLanguage {
            return create(
                languageScopeName,
                grammarRegistry,
                ThemeRegistry.getInstance(),
                autoCompleteEnabled
            )
        }

        fun create(
            languageScopeName: String?,
            grammarRegistry: GrammarRegistry,
            themeRegistry: ThemeRegistry,
            autoCompleteEnabled: Boolean
        ): VCSpaceTMLanguage {
            val grammar = grammarRegistry.findGrammar(languageScopeName)

            requireNotNull(grammar) {
                String.format(
                    "Language with %s scope name not found",
                    grammarRegistry
                )
            }

            val languageConfiguration = grammarRegistry.findLanguageConfiguration(grammar.scopeName)


            return VCSpaceTMLanguage(
                grammar,
                languageConfiguration,
                grammarRegistry,
                themeRegistry,
                autoCompleteEnabled
            )
        }


        fun create(
            grammarDefinition: GrammarDefinition?,
            autoCompleteEnabled: Boolean
        ): VCSpaceTMLanguage {
            return create(grammarDefinition, GrammarRegistry.getInstance(), autoCompleteEnabled)
        }

        fun create(
            grammarDefinition: GrammarDefinition?,
            grammarRegistry: GrammarRegistry,
            autoCompleteEnabled: Boolean
        ): VCSpaceTMLanguage {
            return create(
                grammarDefinition,
                grammarRegistry,
                ThemeRegistry.getInstance(),
                autoCompleteEnabled
            )
        }

        fun create(
            grammarDefinition: GrammarDefinition?,
            grammarRegistry: GrammarRegistry,
            themeRegistry: ThemeRegistry,
            autoCompleteEnabled: Boolean
        ): VCSpaceTMLanguage {
            val grammar = grammarRegistry.loadGrammar(grammarDefinition)

            requireNotNull(grammar) {
                String.format(
                    "Language with %s scope name not found",
                    grammarRegistry
                )
            }

            val languageConfiguration = grammarRegistry.findLanguageConfiguration(grammar.scopeName)

            return VCSpaceTMLanguage(
                grammar,
                languageConfiguration,
                grammarRegistry,
                themeRegistry,
                autoCompleteEnabled
            )
        }
    }
}
